KAFKA-10247: Correctly reset state when task is corrupted

vvcephei opened the pull request:
When we detect a task as corrupted, we need to not only close,
clean up, and recover the local store, but we also need to reset
the consumer to the last committed position so that after recovery
we can start processing from the correct position.

We also need to detect such repaired tasks in the StreamThread
processing loop so they can be re-initialized and recovered, even
when the thread is already in RUNNING.

Also, I fixed a bug in which we would incorrectly always consider
a task as corrupted when it contains both persistent and in-memory
stores.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
KAFKA-10247: Skip processing if task isn't running

abbccdda reviewed (commented):
Thanks for the PR. Not sure I understand the fix here, by throwing ISE, won't the thread just crash?

mjsax commented:
It seem the root cause of the issue is that we don't cleanup internal data structures correctly when closing a task dirty? Should the task not be removed from the list of active tasks instead?

vvcephei commented:
Hey @abbccdda and @mjsax , thanks for taking a look! The PR is not complete yet. I'll let you know when it's ready for review.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -766,6 +769,24 @@ void runOnce() {
+        } else if (builder.latestResetTopicsPattern().matcher(partition.topic()).matches()) {
+            return OffsetResetStrategy.LATEST;
+        } else {
+            if (originalReset == null || (!originalReset.equals("earliest") && !originalReset.equals("latest"))) {
+                return OffsetResetStrategy.EARLIEST;
+            }
```
I honestly couldn't figure out what is the default default default reset strategy... It seems (from the behavior of the test when we first start up) that if there's no strategy set, and no committed offset, then the client starts at the beginning, but the ClientConfig has the default policy as "latest"... What gives?

ableegoldman replied:
Does Streams override the client default..?

mjsax replied:
> Does Streams override the client default..?

Yes. The client default is "latest" but we use "earliest" by default (cf https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L857). Of course, users can also change the default via StreamsConfig.

Note that the consumer client can only apply a single strategy to all topics it subscribed to. Hence, if all topics use the same reset policy, we can rely on the consumer configures policy. However, if users specify different reset policies in their code via `Consumed` for individual topics, the consumer is re-configured to use "none" (cf. https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L362-L366) and we do a manual seekToBeginning/seekToEnd according to the user define strategy for the corresponding topic (cf https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L762-L764) because we need to make a per-topic decision that the consumer cannot make for us.

vvcephei replied:
Ah, thanks for the explanation, @mjsax !

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -193,6 +197,28 @@ private void closeAndRevive(final Map<Task, Collection<TopicPartition>> taskWith
                 log.error("Error suspending corrupted task {} ", task.id(), swallow);
             }
             task.closeDirty();
+            // Pause so we won't poll any more records for this task until it has been re-initialized
+            // Note, closeDirty already clears the partitiongroup for the task.
+            mainConsumer().pause(task.inputPartitions());
+            final Map<TopicPartition, OffsetAndMetadata> committed = mainConsumer().committed(task.inputPartitions());
+            for (final TopicPartition topicPartition : task.inputPartitions()) {
+                final OffsetAndMetadata offsetAndMetadata = committed.get(topicPartition);
+                if (offsetAndMetadata == null) {
+                    final OffsetResetStrategy strategy = resetStrategy.apply(topicPartition);
+                    switch (strategy) {
+                        case EARLIEST:
+                            mainConsumer().seekToBeginning(Collections.singleton(topicPartition));
+                            break;
+                        case LATEST:
+                            mainConsumer().seekToBeginning(Collections.singleton(topicPartition));
+                            break;
+                        default:
+                            throw new IllegalArgumentException("Unexpected reset strategy: " + strategy);
+                    }
+                } else {
+                    mainConsumer().seek(topicPartition, offsetAndMetadata);
+                }
+            }
```
This might be the worst thing I've ever proposed for AK... I can't figure out a better way to just "reset" the offset.

mjsax replied:
Why do you think is bad? That is just how the API works... Cf https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L769-L802 that does the same thing.

What make we wonder, if we can share common code for both cases?

vvcephei replied:
I think we can, too. I'll give it a shot, since it doesn't seem crazy to you :)

vvcephei commented:
Ok, @abbccdda and @mjsax , I think this is ready for a review now. I'm not very happy with the logic to reset to the last committed offset. Can you think of a better way to do it?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -193,6 +197,28 @@ private void closeAndRevive(final Map<Task, Collection<TopicPartition>> taskWith
+                            mainConsumer().seekToBeginning(Collections.singleton(topicPartition));
+                            break;
+                        case LATEST:
+                            mainConsumer().seekToBeginning(Collections.singleton(topicPartition));
```
Should this be `seekToEnd` ?

vvcephei replied:
ah... yes.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -95,7 +96,7 @@
      *          |      +-----+-------+      |
      *          |            |              |
-     *          |            |              |
+     *          |            |--------------+
```
What's up with this?

vvcephei replied:
I meant to call this out, too. In order to get the task re-initialized after a corruption recovery, we could set the state back to PartitionsAssigned, but I felt that would be confusing in the logs. Instead, I added an extra condition that we'll initialize tasks if there are any that need initialization, even though the thread may already be in running, which means we can have a self-transition from running to running.

mjsax replied:
Where does this self-transition happen exactly? And could/should we detect this case and not call `setState()` for this case instead of allowing the transition?

mjsax replied:
\cc @vvcephei

vvcephei replied:
Ah, sorry, I didn't see this follow-up question.

It's here: https://github.com/apache/kafka/blob/568dceb5f61c71428432d23d947c5f9b29fb7bfb/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L657

I thought about it, but it seemed to be in violation of the whole idea of having a specified state machine. I.e., saying "`if not running, setState(running)`" is literally the same thing as saying that self-transitions are allowed. So why not just say that self-transitions are allowed (explicitly instead of implicitly)?

Anyway, that's what I was thinking; no guarantee that it makes sense ;)

mjsax replied:
I agree with the sentiment. However, our overall patterns is to disallowed self-transitions, ie, this pattern is applied throughout all other code. IMHO, I might be better to stick with one pattern. Thus, if we think we should allow self-transitions, we might want to do it for all states and update the code throughout accordingly. Begin in a "mixed mode" seems suboptimal.

Of course this would be follow up work. And frankly I am not sure if it would buy us much at this point?

ableegoldman replied:
FWIW, we actually do follow this pattern already, at least with respect to StreamThread state transitions. I'd rather be consistent with the other StreamThread state changes and inconsistent with other places in the code (eg the Task state changes), especially if we all agree this seems like the sensible approach to the fsm

mjsax replied:
Thanks for chiming in. Seems I did not separate thread state and task state transition. Sorry for the confusion.

vvcephei replied:
No worries, thanks for the review @mjsax . Now that I'm thinking about it, it does seem like Task is _allowing_ self-transitions, rather than _disallowing_ them, so we should probably add the self-transitions to the Task state machine. But maybe in a follow-up.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -137,7 +138,7 @@
         PARTITIONS_REVOKED(2, 3, 5),      // 2
         PARTITIONS_ASSIGNED(2, 3, 4, 5),  // 3
-        RUNNING(2, 3, 5),                 // 4
+        RUNNING(2, 3, 4, 5),              // 4
```
Why do we need this?

vvcephei replied:
Sorry, I meant to explain it, but forgot. I've explained it here: https://github.com/apache/kafka/pull/8994#discussion_r451938891

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -766,6 +769,24 @@ void runOnce() {
         return records;
     }
 
+    private OffsetResetStrategy getResetStrategy(final TopicPartition partition) {
```
Wondering if we should reuse this method within `StreamThread#resetInvalidOffsets`?

vvcephei replied:
Yeah, I think we should.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1140,4 +1166,12 @@ public static void executeAndMaybeSwallow(final boolean clean,
+        return tasks().values().stream().anyMatch(Task::preRunning);
+    }
+
+    public void setResetStrategy(final Function<TopicPartition, OffsetResetStrategy> resetStrategy) {
```
I think the name of the method is not ideal. We don't set the strategy, but we set a function that can compute the strategy for a partitions. Needed to go forth and back on the PR to understand how it work, and assume the method name had its part in confusing me.

vvcephei replied:
Sorry about that. I was only thinking "set a field named 'reset strategy'", but now I see that it sounds more like what you said. I'll rename it.

mjsax reviewed (commented):
Did not review the test code yet.

vvcephei commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -504,7 +504,7 @@ private void resetOffsetPosition(TopicPartition tp) {
             offset = beginningOffsets.get(tp);
             if (offset == null)
-                throw new IllegalStateException("MockConsumer didn't have beginning offset specified, but tried to seek to beginning");
+                throw new IllegalStateException("MockConsumer didn't have beginning offset for " + tp + " specified, but tried to seek to beginning");
```
Just a quality-of-life improvement.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -215,7 +215,7 @@ public StateStore getGlobalStore(final String name) {
 
     // package-private for test only
-    void initializeStoreOffsetsFromCheckpoint(final boolean storeDirIsEmpty) {
+    void initializeStoreOffsetsFromCheckpoint(final boolean taskDirIsEmpty) {
```
Was deceptively named.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -232,10 +232,16 @@ void initializeStoreOffsetsFromCheckpoint(final boolean storeDirIsEmpty) {
+                        // two stores in a task, and one is correctly written and checkpointed, while the other is
+                        // neither written nor checkpointed, we _could_ correctly load the first and recover the second
+                        // but instead we'll consider the whole task corrupted and discard the first and recover both.
+                        if (store.stateStore.persistent() && eosEnabled && !taskDirIsEmpty) {
```
Bugfix: we shouldn't call this task corrupted for not having a checkpoint of a non-persistent store.

vvcephei replied:
I'm not sure how this came up now for me. I don't think it was related to any changes in this PR, but the SuppressionDurabilityIntegrationTest fails every time without it.

ableegoldman replied:
I like my fix better :P 

But seriously, no need to block this PR on mine if it's suddenly causing tests to fail. Mysterious..

vvcephei replied:
Oh, I put this comment on the wrong line:

> FYI, this is also fixed (better) in #8996

In other words, I agree, I like your fix better as well.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -462,7 +468,7 @@ public void flush() {
     @Override
     public void close() throws ProcessorStateException {
-        log.debug("Closing its state manager and all the registered state stores: {}", stores);
+        log.info("Closing its state manager and all the registered state stores: {}", stores);
```
```suggestion
        log.debug("Closing its state manager and all the registered state stores: {}", stores);
```

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -648,7 +650,7 @@ void runOnce() {
         // only try to initialize the assigned tasks
         // if the state is still in PARTITION_ASSIGNED after the poll call
-        if (state == State.PARTITIONS_ASSIGNED) {
+        if (state == State.PARTITIONS_ASSIGNED || taskManager.hasPreRunningTasks()) {
```
Bugfix: If a task has been revived, then it needs to be initialized and restored, even if the thread state is already RUNNING.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -768,9 +770,30 @@ void runOnce() {
 
     private void resetInvalidOffsets(final InvalidOffsetException e) {
         final Set<TopicPartition> partitions = e.partitions();
+        final Set<TopicPartition> notReset = resetOffsets(partitions);
```
Refactored to share the new resetOffsets method.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -193,6 +196,26 @@ private void closeAndRevive(final Map<Task, Collection<TopicPartition>> taskWith
                 log.error("Error suspending corrupted task {} ", task.id(), swallow);
             }
             task.closeDirty();
+            if (task.isActive()) {
```
Shouldn't try to reset the main consumer at all if this is only a standby.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -193,6 +196,26 @@ private void closeAndRevive(final Map<Task, Collection<TopicPartition>> taskWith
+                // Pause so we won't poll any more records for this task until it has been re-initialized
+                // Note, closeDirty already clears the partitiongroup for the task.
+                final Set<TopicPartition> currentAssignment = mainConsumer().assignment();
+                final Set<TopicPartition> assignedToPauseAndReset =
+                    Utils.intersection(HashSet::new, currentAssignment, task.inputPartitions());
```
I don't fully understand how this could happen, but I saw it in several tests, that an active task isn't assigned the input TopicPartition for itself. Whether or not it can happen with a real consumer, or only in tests with the MockConsumer, doesn't really seem important. If the topic isn't assigned, we certainly don't need to reset it.

vvcephei replied:
FYI, this is also fixed (better) in https://github.com/apache/kafka/pull/8996

mjsax replied:
I agree that it should never happen. Maybe it's a test setup issue? Should we be worries to mask a potential bug by writing the code like this? Or should we rather fail hard (if we assume it would indicate a bug) and make sure the update the tests?

ableegoldman replied:
Could this happen due to some race condition with a deleted regex topic? I assume that's not happening in your tests, just wondering if this check actually is necessary or if it would only be masking a bug as Matthias said.

I'm _pretty_ sure we should always go through the process of updating the task's input partitions immediately after updating the consumer's assignment, so we would never find them to be out of sync.  But can the consumer assignment be updated outside of a rebalance, eg based on metadata refresh?

vvcephei replied:
Thanks all, it may well have just been a problem with the test setup, but @ableegoldman is making me nervous that there may be some edge cases where it's expected in production code.

Since we would never have had a dependency on the task inputs actually being assigned here, and now we do, I'm concerned that if we try to be strict right now, we may be introducing an IllegalStateException. Normally, this wouldn't be a problem, but this PR is a last-minute hotfix before a release, so I'm feeling cautious.

Since we seem to agree that the code is correct, although it may not be strict, I've added a warning log instead of increasing strictness.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -566,8 +567,20 @@ public void shouldReviveCorruptTasks() {
         expectLastCall().anyTimes();
 
+        expect(consumer.assignment()).andReturn(taskId00Partitions);
+        consumer.pause(taskId00Partitions);
+        expectLastCall();
+        final OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(0L);
+        expect(consumer.committed(taskId00Partitions)).andReturn(singletonMap(t1p0, offsetAndMetadata));
+        consumer.seek(t1p0, offsetAndMetadata);
+        expectLastCall();
+        consumer.seekToBeginning(emptySet());
+        expectLastCall();
         replay(activeTaskCreator, topologyBuilder, consumer, changeLogReader);
-
+        taskManager.setPartitionResetter(tp -> {
+            assertThat(tp, is(empty()));
+            return emptySet();
+        });
```
This all amounts to checking that we really reset the consumer to the last committed position.

vvcephei reviewed (commented):
Hey @abbccdda @mjsax @ableegoldman ,

I had some failing tests after the last refactor, and chasing them down led me to make some changes, so I've gone over the PR again with a fresh self-review. Can you take another look?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -232,10 +232,16 @@ void initializeStoreOffsetsFromCheckpoint(final boolean storeDirIsEmpty) {
+                        // directory is nonempty, only if there are any directories for any stores. So if there are
+                        // two stores in a task, and one is correctly written and checkpointed, while the other is
+                        // neither written nor checkpointed, we _could_ correctly load the first and recover the second
+                        // but instead we'll consider the whole task corrupted and discard the first and recover both.
```
Sound like something we should fix. Can you file a ticket?

vvcephei replied:
IIRC, @guozhangwang tried to include this during the February refactor, but it's harder to get right than it sounds. Still, it would be very nice to have it, and I agree it would be good to file a ticket.

vvcephei replied:
https://issues.apache.org/jira/browse/KAFKA-10260

Commit:
rebase on trunk

vvcephei commented:
Hi @abbccdda @ableegoldman @mjsax ,

I've rebased on trunk to resolve conflicts with Sophie's parallel fix. I also dropped a few unnecessary side-fixes. Do you mind giving this a final pass?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -768,9 +770,30 @@ void runOnce() {
+            final String notResetString =
+                notReset.stream()
+                        .map(tp -> "topic " + tp.topic() + "(partition " + tp.partition() + ")")
+                        .collect(Collectors.joining(","));
```
Nit: As reset policy is set on a per topic basis, it's sufficient to list the topic names -- it does not add value if we list the partitions, because all assigned partitions would be affected anyway.

vvcephei replied:
Interesting; good point. I just copied over the original message. Do you think I should drop the partition?

mjsax replied:
I think it would be good. Seems the original message should not have contained the partition to begin with. Even if for the original one, it's less severe because we only consider a single topic-partition in the error message.

For this new case, we get multiple topic-partitions and the error message might become quite large? Also, if multiple task of the same sub-topology are assigned, the same topic name might appear multiple times for different partitions? Might be best to only add a `Set` or topic names to the error message?

vvcephei replied:
Ok, I've done this.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -193,6 +196,26 @@ private void closeAndRevive(final Map<Task, Collection<TopicPartition>> taskWith
+                }
+                final Set<TopicPartition> remainder = resetter.apply(assignedToPauseAndReset);
+                // If anything didn't have a configured policy, reset to beginning
+                mainConsumer().seekToBeginning(remainder);
```
I think we should fail for this case, because if user configures "none" they request that we fail if we loose track of the offset.

vvcephei replied:
Thanks. I was confused by the behaviour of the StreamThreadTest. I think there was something funny in the StreamThread factory, which just never surfaced before. I'll call it out for your review.

Commit:
cr comments

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -359,11 +359,8 @@ public static StreamThread create(final InternalTopologyBuilder builder,
         consumerConfigs.put(StreamsConfig.InternalConfig.ASSIGNMENT_ERROR_CODE, assignmentErrorCode);
         final AtomicLong nextScheduledRebalanceMs = new AtomicLong(Long.MAX_VALUE);
         consumerConfigs.put(StreamsConfig.InternalConfig.NEXT_SCHEDULED_REBALANCE_MS, nextScheduledRebalanceMs);
-        String originalReset = null;
-        if (!builder.latestResetTopicsPattern().pattern().equals("") || !builder.earliestResetTopicsPattern().pattern().equals("")) {
-            originalReset = (String) consumerConfigs.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG);
-            consumerConfigs.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "none");
-        }
```
I had to change this to get the StreamThread test to actually use the configured `ConsumerConfig.AUTO_OFFSET_RESET_CONFIG := earliest`

Reading the conditional, it doesn't make any sense to me, but it's been in the codebase for a long time, so I'm doubting myself. It seems to say that we will only use the provided client configuration if there **is** an override, but it seems like it should have been "if there is **not** an override".

Regardless, the "originalReset" is only used as a fallback _after_ we apply the builder reset patterns, so I don't see why we should leave it null in any case.

ableegoldman replied:
This makes my head hurt..

mjsax replied:
If users specify an in-code overwrite via `Consumed` the user might do this only for _some_ topics. For all other topics, the configures reset policy should be used. Hence, we just "remember" the default config in `originalReset` and set the consumer config to "none".

If there is no in-code overwrite, we don't need to remember the original strategy because all topics use the same strategy and thus we rely on the consumer anyway to do the reset for us.

With this change, we now _always_ set the reset policy to "none". What is still possible of course, but it implies, we never rely on the consumer any longer to do any reset and we always to it manually. (This might be "cleaner" as it might be easier to reason about.) -- But the old code was correct, too.

mjsax replied:
I guess the main reason for this old design was, that user see their config when `ConsumerConfig` is logged. -- Now, we would always log "none" independent what the users sets, and this might be confusing.

I think we should update the docs accordingly and call it out.

vvcephei replied:
Aha! Thanks, @mjsax. This makes sense. I hadn't considered that the consumer never throws `InvalidOffsetException` if there's a valid policy configured, so the prior logic only needs the fallback when there's an override.

Now that we also use the resetting logic for our own "manual" reset when handling TaskCorruptedException, we do need to capture the Consumer config value, but we can still enforce that if there are any overrides in the topology, we set the consumer to "none". Just updated the code to do this.

Commit:
cr comments

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -767,38 +766,65 @@ void runOnce() {
     }
 
     private void resetInvalidOffsets(final InvalidOffsetException e) {
```
This method is a one-liner now and is only called in a single place IIRC. Maybe better to remove the method and embed the call?

vvcephei replied:
Good catch! I'll inline it.

Commit:
CR comment

Commit:
partially revert originalReset logic

Commit:
inline the reset functions

vvcephei commented:
Retest this please

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -760,44 +763,62 @@ void runOnce() {
-        if (!seekToEnd.isEmpty()) {
-            mainConsumer.seekToEnd(seekToEnd);
+        if (notReset.isEmpty()) {
+            if (!seekToBeginning.isEmpty()) {
```
We we need this `isEmpty` check? What happens is we blindly pass an empty set into `seekToBeginning`? -- Similar for `seekToEnd()` below?

vvcephei replied:
Huh, I didn't wonder that before, but ... dear god. From the javadoc on KafkaConsumer:
> If no partitions are provided, seek to the first offset for all of the currently assigned partitions.

What a dangerous API!

vvcephei replied:
Just verified that we never call it unguarded by a non-empty check.

mjsax replied:
Awesome! Thanks for double checking!

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -193,6 +195,34 @@ private void closeAndRevive(final Map<Task, Collection<TopicPartition>> taskWith
+                    );
+                }
+
+                mainConsumer().pause(assignedToPauseAndReset);
```
For my own education: when does the actual `resume()` happen (and are we sure those partitions are resumed later?)

vvcephei replied:
Good question!
It goes through the "restoration" phase, at the end of which, we resume the main consumer's partitions.

In TaskManager:
```
        if (allRunning) {
            // we can call resume multiple times since it is idempotent.
            mainConsumer.resume(mainConsumer.assignment());
        }
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1177,6 +1180,109 @@ public void shouldNotCloseTaskAndRemoveFromTaskManagerIfProducerGotFencedInCommi
+            ),
+            "proc"
+        );
+        internalTopologyBuilder.addSink("out", "output", null, null, null, "proc");
```
Do we need the sink?

vvcephei replied:
Uh, no. It was just copied with the rest of the test. Sorry about that.

vvcephei replied:
fixed this, since I needed more commits anyway.

mjsax commented:
Just two last comments/questions. Overall LGTM.

mjsax reviewed (approved):
Feel free to merge after Jenkins passed.

vvcephei commented:
Huh, one of those tests hit:
```
Exception in thread "stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1" java.util.ConcurrentModificationException
	at java.util.TreeMap$PrivateEntryIterator.nextEntry(TreeMap.java:1211)
	at java.util.TreeMap$ValueIterator.next(TreeMap.java:1256)
	at java.util.AbstractCollection.toArray(AbstractCollection.java:141)
	at java.util.LinkedList.addAll(LinkedList.java:408)
	at java.util.LinkedList.addAll(LinkedList.java:387)
	at java.util.LinkedList.<init>(LinkedList.java:119)
	at org.apache.kafka.streams.processor.internals.TaskManager.needsInitializationOrRestoration(TaskManager.java:1175)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:654)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:551)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:510)
```

Commit:
fix race condition

Commit:
remove unnecessary node from test

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -648,7 +651,9 @@ void runOnce() {
         // only try to initialize the assigned tasks
         // if the state is still in PARTITION_ASSIGNED after the poll call
-        if (state == State.PARTITIONS_ASSIGNED) {
+        if (state == State.PARTITIONS_ASSIGNED
+            || state == State.RUNNING && taskManager.needsInitializationOrRestoration()) {
```
The flaky test was related to us going into this block from other states. I finally got a clue when one of the tests failed on "invalid transition from PARTITIONS_REVOKED to RUNNING". I'm not sure how, exactly, but I think the shutdown test that failed on ConcurrentModificationException was also related, probably due to the test invoking the handleAssignment/Revocation/Lost methods from a different thread (which can normally never happen).

Anyway, my prior code only intended to add the _self_ transition, but failed to make sure we were actually in a self-transition. It's fixed now.

mjsax replied:
Thanks. The fix makes sense to me.

vvcephei reviewed (commented):
Unfortunately (that failure having been a multithreaded bug), I'm not 100% sure this was it, but after this change, I was able to run all the tests in this class 3,000 times without failure, so it _seems_ like I got it.

The fix (which I call out below) was an obvious bug anyway.

mjsax commented:
Merged to `trunk` and cherry-picked to `2.6`.

vvcephei commented:
Thanks, @mjsax !
